### PR TITLE
feat: Versionable Merkle metadata

### PR DESCRIPTION
**Pull request by bvrooman**
Related issues:
- #1552

### CODE REVIEW

File: crates/storage/src/blueprint/sparse.rs [hunk 1]
```diff
@@ -377,9 +380,9 @@ where
         let prev_metadata: Cow<SparseMerkleMetadata> = storage
             .storage::<Metadata>()
             .get(primary_key)?
-            .unwrap_or_default();
+            .unwrap_or(Cow::Owned(SparseMerkleMetadataV1::default().into()));
```
Comment (xgreenx) [inline crates/storage/src/blueprint/sparse.rs:383]: You can implement `Default` for the enum; it is okay to do since we can use the latest(primary) variant of the enum if there is no entry in the storage.

File: crates/storage/src/blueprint/sparse.rs [hunk 2]
```diff
@@ -152,7 +155,7 @@ where
                 storage.storage::<Metadata>().remove(primary_key)?;
             } else {
                 // Generate new metadata for the updated tree
-                let metadata = SparseMerkleMetadata { root };
+                let metadata = SparseMerkleMetadataV1 { root }.into();
```
Comment (xgreenx) [inline crates/storage/src/blueprint/sparse.rs:158]: Maybe it is okay to add a new constructor that creates `SparseMerkleMetadata` from the root.

File: crates/fuel-core/src/database/block.rs [hunk 3]
```diff
@@ -115,18 +115,18 @@ impl StorageMutate<FuelBlocks> for Database {
             .next()
             .transpose()?
             .map(|(_, metadata)| metadata)
-            .unwrap_or_default();
```
Comment (xgreenx) [inline crates/fuel-core/src/database/block.rs:118]: You can revert this change=)

File: crates/fuel-core/src/database/block.rs [hunk 4]
```diff
@@ -115,18 +115,18 @@ impl StorageMutate<FuelBlocks> for Database {
...
-            MerkleTree::load(storage, prev_metadata.version)
+            MerkleTree::load(storage, prev_metadata.version())
                 .map_err(|err| StorageError::Other(anyhow::anyhow!(err)))?;
         tree.push(block_id.as_slice())?;
 
         // Generate new metadata for the updated tree
         let version = tree.leaves_count();
         let root = tree.root();
-        let metadata = DenseMerkleMetadata { version, root };
+        let metadata = DenseMerkleMetadataV1 { version, root }.into();
```
Comment (xgreenx) [inline crates/fuel-core/src/database/block.rs:129]: The same here, we can have a basic constructor since these fields will be used in most cases

File: CHANGELOG.md [hunk 5]
```diff
@@ -25,6 +25,7 @@ Description of the upcoming release here.
 - [#1601](https://github.com/FuelLabs/fuel-core/pull/1601): Fix formatting in docs and check that `cargo doc` passes in the CI.
 
 #### Breaking
+- [#16239](https://github.com/FuelLabs/fuel-core/pull/1639): Make Merkle metadata, i.e. `SparseMerkleMetadata` and `DenseMerkleMetadata` type version-able enums
```
Comment (matt-user) [inline CHANGELOG.md:28]:
```suggestion
- [#1639](https://github.com/FuelLabs/fuel-core/pull/1639): Make Merkle metadata, i.e. `SparseMerkleMetadata` and `DenseMerkleMetadata` type version-able enums
```

**Review by xgreenx — APPROVED** [review]
